src/hotspot/share/gc/g1/g1FullGCCompactionPoint.cpp

```diff
@@ -130,3 +131,86 @@ void G1FullGCCompactionPoint::remove_at_or_above(uint bottom) {
   assert(start_index >= 0, "Should have at least one region");
   _compaction_regions->trunc_to(start_index);
 }
+
+void G1FullGCCompactionPoint::add_humongous(HeapRegion* hr) {
+  assert(hr->is_starts_humongous(), "Sanity!");
+
+  _collector->add_humongous_region(hr);
+
+  G1CollectedHeap* g1h = G1CollectedHeap::heap();
+  do {
+    add(hr);
+    _collector->update_from_skip_compacting_to_compacting(hr->hrm_index());
+    hr = g1h->next_region_in_humongous(hr);
+  } while (hr != nullptr);
+}
+
+uint G1FullGCCompactionPoint::forward_humongous(HeapRegion* hr) {
+  assert(hr->is_starts_humongous(), "Sanity!");
+
+  oop obj = cast_to_oop(hr->bottom());
+  size_t obj_size = obj->size();
+  int num_regions = (int) G1CollectedHeap::humongous_obj_size_in_regions(obj_size);
+
+  if (!has_regions()) {
+    return num_regions;
+  }
+
+  // Find contiguous compaction target regions for the humongous object
+  Pair<uint, uint> range = find_contiguous_before(hr, num_regions);
```

Comment:
(I'm suprised that MSVC does not complain here passing an int to the uint parameter)